go-controller/pkg/config/cni.go

```diff
@@ -71,29 +72,86 @@ func WriteCNIConfig() error {
 
 // ParseNetConf parses config in NAD spec
 func ParseNetConf(bytes []byte) (*ovncnitypes.NetConf, error) {
+	netconf, isRecoverable, err := parseNetConfList(bytes)
+	if err != nil {
+		if isRecoverable {
+			netconf, _, err = parseNetConfSingle(bytes)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	}
+
+	if netconf.Topology == "" {
+		// NAD of default network
+		netconf.Name = ovntypes.DefaultNetworkName
+	}
+
+	return netconf, nil
+}
+
+func parseNetConfSingle(bytes []byte) (*ovncnitypes.NetConf, bool, error) {
 	netconf := &ovncnitypes.NetConf{MTU: Default.MTU}
 	err := json.Unmarshal(bytes, &netconf)
 	if err != nil {
-		return nil, err
+		return nil, true, err
 	}
+
 	// skip non-OVN NAD
 	if netconf.Type != "ovn-k8s-cni-overlay" {
-		return nil, ErrorAttachDefNotOvnManaged
+		return nil, true, ErrorAttachDefNotOvnManaged
 	}
-	if netconf.Topology == "" {
-		// NAD of default network
-		netconf.Name = ovntypes.DefaultNetworkName
-	} else {
+
+	err = validateNetConf(netconf)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return netconf, true, nil
+}
+
+func parseNetConfList(bytes []byte) (*ovncnitypes.NetConf, bool, error) {
+	netconf := &ovncnitypes.NetConf{MTU: Default.MTU}
+	netConfList := &ovncnitypes.NetConfList{Plugins: []*ovncnitypes.NetConf{netconf}}
+	err := json.Unmarshal(bytes, &netConfList)
+	if err != nil {
+		return nil, true, err
+	}
+
+	if len(netConfList.Plugins) != 1 {
+		return nil, false, ErrorPluginsListSize
+	}
+
+	// skip non-OVN NAD
+	if netconf.Type != "ovn-k8s-cni-overlay" {
+		return nil, true, ErrorAttachDefNotOvnManaged
```

Comment:
why is this one recoverable ?

Reply:
same as https://github.com/ovn-org/ovn-kubernetes/pull/3844#discussion_r1299930998 isnt it ?

Reply:
Why did you feel the need of having the functions return the same type of arguments ? Simply put, `parseNetConfList` requires a netconf parse fallback, but `parseNetConfSingle` does not.

Making these functions have the same return arguments is wrong, and misleads the code reader - we're essentially saying there's something to fallback to when `parseNetConfSingle` fails. Unless I'm missing something, there just isn't.

Reply:
sumup here https://github.com/ovn-org/ovn-kubernetes/pull/3844#discussion_r1299973223
will change